### ISSUE TITLE
implement lightweight CRAB distribution as part of CMSSW distribution

### ISSUE THREAD

**Issue by belforte** [dmwm/CRABClient#4844]
As per kickoff chat on Oct 30 . Participant were:
Steano, Marco Mascheroni, Leonardo, Bockjoo, Shahzad, David Lange.

Notes from the chat from Stefano. If anything wrong is written, my fault.
These are also in the Minutes in Indico:
https://indico.cern.ch/event/859940/

Thanks Leonardo for taking responsibility for this from now on.
 
Crab Client is only python 2.7 with only dependance from pycurl.
Crab developers do not want to bring any binary.
 
We agreed to try to move CRAB Client into an external distributed via CMSSW
 distribution release. The idea is that CRAB Client and CMSSW can be
 executed in the same environment. Many users have python scripts where
 they call CRAB python API while manipulating CMSSW configuration, we
 do not want to break those.
 
CrabClient will not be part of the release, so CRAB version stays the same whichever
  CMSSW version people use. CRAB version will be updated when the distribution is
  updated, i.e. at the time of any new release build (usually every two weeks).
The distribution update will push out whatever CRAB tag we set in the spec file
 (in CMSSW repository) via a PullRequest. The new version will be available in the
 IB in the meanwhile.
 
Given the above constrains we will have two versions of CRAB Client in the
 CMSSW distribution: pre and prod, or beta an prod, or whatever... which we
 will update via separate PR. So that we can update the PRE, make sure it
 works, and only when sure, update production. If we push a bug in production
 we may need to wait two weeks for the fix !!
 
If this works, Bockjoo can stop the current machinery for CRAB and will not need
to do anything.
 
We will most likely need to change/add some setup script in CRAB, details need
 to be figured out.
 
Marco Mascheroni is available to help. Shahzad is willing to help us write
 the needed spec file.
 
Leonardo is in charge of this effort and will communicate progress or problems.
 
Next week Marco will be at CERN and Leonardo, Shahzad and Marco will get
 toghether to get things started.,
 
We do not need to have python3 compatible version. But of course in time
we may want to.

**Comment by lecriste** [comment]
Update from November 6th:
@smuzaffar will incorporate the [current spec file](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/crabclient.spec) in CMSSW and try to build RPMs. A `crab-dev` command will be available upon issuing `cmsenv` in order to perform the needed test. When we are happy it will be renamed `crab` and we can stop sourcing the current file in `/cvmfs/cms.cern.ch/crab3/`.
If this all works, the procedure to release a new client version will not change (apart from the spec file location).
A `crab-pre` command will also be available, pointing to the preprod RPM.
Need to make sure that the use case where the user runs a python script to build both CRAB config and CMSSW PSet as python objects and submit on the fly via CRAB client API is preserved.

**Comment by smuzaffar** [comment]
@lecriste @belforte please see https://github.com/cms-sw/cmsdist/pull/5356
this is my first attepmt to include crab client in cmssw software stack. My initial understanding was that there is few simple python scripts for crab client which only depend on pycurl but it turned out (trying to build https://github.com/cms-sw/cmsdist/blob/comp_gcc630/crabclient.spec ) that it depends on many python and DMWM packages. For now I have simplified the build recipe ( https://github.com/cms-sw/cmsdist/pull/5356/files#diff-7d6e1f3d0fe3aaef69481cabb9985504 ) to just get directly the DMWM tools and only depend on `py2-pycurl py2-python-cjson`

Once the cmsdist PR is merged then new crab wrapper scripts ( https://github.com/cms-sw/cms-common/blob/master/common/crab-prod {dev,pre} ) will be available under `/cvmfs/cms-ib.cern.ch/week1/common` and you should be able to run `/cvmfs/cms-ib.cern.ch/week1/common/crab-prod`. I have tested this locally and was able to run `sumbit` and `query` commands. 

@belforte , this only solves the command-line usage of crab-client. The `CRAB python API` is still need more work.

**Comment by smuzaffar** [comment]
By the way, as mentioned during our meeting. Can we change crab so that the crab wrapper script always unset cmssw env (i.e by running `eval scram unset -sh`) and crab python sets it back when ever it has to deal with cmssw?

**Comment by belforte** [comment]
hmmm... how would crab find python and pycurl then ?
Sorry if I did not understand things in the vidyo meeting. Was this discussed with Leo and Marco ?

**Comment by belforte** [comment]
Sorry for poor communication. There can be a lot of python files involved,
but AFAIK the only binary dependence is pycurl. We do not use cjson anymore,
I removed references from the code but did not think about the spec file,
should I ? Could Leo ?
If further simplifications are needed, we will look into it.

Can you elaborate on the problem with CRAB python API ?
AFAIU (and I can be wrong which why I asked Marco's opinion) all that
CRAB needs after cmsenv is to put a bunch of stuff on $PYTHONPATH
And nothing in that stuff should have same names, and hence possibly
conflict, with CMSSW. What am I missing ?

**Comment by smuzaffar** [comment]
crab wrapper  script finds the latest version of `crab` ( https://github.com/cms-sw/cms-common/blob/master/common/crab-prod#L11-L15 ) and source its `init.sh` to set env for crab. This should set the correct path to python, pycurl and any other packages needed for crab itself.  

I am not sure how exactly crab uses cmssw? If it is just running some cmssw commands to dump cmssw configuration then unsetting cmssw env in crab wrapper and setting it again (in sub-shell/process) before calling cmssw will work. But if crab imports cmssw python configuration then it will not work.

**Comment by smuzaffar** [comment]
problem with PYTHON API is that when someone imports `CRABClient` from their python scripts then this means python interpreter has already been started and we can not change python version for crab and might not be able to unset cmssw env

**Comment by belforte** [comment]
one thing at a time
1. **crab wrapper** :  I hope we can pitch it (since I never understood the intricated nesting done in there) - was hoping for @lecriste and @mmascher to take the lead here. Can you guys ? 
2. **python API**: python calls CRAB/CMSSW are mixed in, but the only place where we need to do scram unsetenv is in a forked subprocess before calling dreaded grid clients like voms-proxy* and gfal* which have their own binary stack that we do no want to know about. That forked subprocess does not need CRAB env. either.
3. **python version**:  that's the only thing I really worry about, since at some point we may have a python 3 CRAB used with an old CMSSW release which has python 2.7 inside. Can we imagine to write CRAB Client so that it "always work"  ? Or will we need to support two versions ? This would be a problem also with current setup. Mayb/hopefully  I am simply confused.

**Comment by belforte** [comment]
One problem here is that there's a large distance between what's needed to make CRABClient run, and what's declared in current spec files, and what's defined in current setup script.

FWIW I could run crab submit, crab status and crab checkwrite (hence I suspect everyhing) after:
1.  `cmsenv`
2. `export PATH=/afs/cern.ch/user/b/belforte/WORK/CRAB3/CRABClient/bin:$PATH`
3. `export PYTHONPATH=<some stuff from github>:$PYTHONPATH`
4. where `<some stuff from github>` means
 * CRABClient (of course)
 * 3 files from CRABServer (including a `PandaServerInterface.py` which we should better deprecate)
 * WMCore (where we only use a very limited subset of course)
 * DBS/Client and DBS/PycurlClient

some additional definitions will be needed to e.g. make command autocompletion work, but  I think that's what I mean when I say "CRAB Client is just a set of python script + pycurl"

question is: what's the simplest way to configure things so that thin can easily setup and run.

**Comment by belforte** [comment]
The "python version" question needs to be clarified.
How far back in python do we need to be compatible ?
e.g. we have code (which I'd like to get rid of) to protect in case of python 2.4, see https://github.com/dmwm/CRABClient/issues/4856#issuecomment-554460957

**Comment by smuzaffar** [comment]
I do not think we have any active cmssw release where we have python 2.4. CMSSW_5_3 uses python 2.6

**Comment by smuzaffar** [comment]
`3 python version:`
  If you have two different versions of CRAB for python2 and 3 then yes it will hard to setup the env. How hard it is to have crab client (and its dependencies e.g CarbServer, DBS and WMCore) to  support both version of python?

**Comment by belforte** [comment]
As long as it possible, should not be hard !
Would be good to know @amaltaro 's plans on this.

**Comment by amaltaro** [comment]
Our goal for WMCore is to have it supporting both python2.7 and python3.x, at least until we can fully migrate it to python3.

**Comment by belforte** [comment]
Great. I guess that if you ever decide to make http request stuff (pycurl wrappers)
incompatible with python 2.7, we can fork a frozen version inside CrabClient
repository. I guess we need to make our core py2/3 compatible, which for
some reasons was not fully done, iteration on dictionaries key/val is
still in py27 format. We'll see.

**Comment by smuzaffar** [comment]
Crab client is now available `/cvmfs/cms-ib.cern.ch/week1/common/crab-{prod|dev|pre}`. From lxplus, you should be able to run it without setting any env (you need cmssw env for `crab-prod sumbit` command only)

- In order to make it work with any version of CMSSW (for now only slc7 based releases), we have to make sure that it works with system python. Using python from out externals causes tools like gfal to fail as these tools explicitly use system python. 

- Currently Crab client spec ( https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_11_1_X/master/crab-build.file ) only requires `curl` (and its dependencies) from cms externals. This is needed to have curl build with OpenSSL backend (system curl is build with NSS and does not recognize CERN CA certificates)

- CrabClient still depends on `cjson` via `dbs` [a]. For now I have included https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_11_1_X/master/cjson.py.file in the crab lib directory and crab wrapper script ( https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_11_1_X/master/_crab-startup.file#L36-L38 ) includes it if cjson is not available via system.

- pycurl should comes from system too (it is available on both slc6 and slc7).

Can you please give it a try and see if you can run most of crab commands?

[a]
```
Traceback (most recent call last):
  File "crab/3.3.1912.rc1-cms/bin/crab", line 157, in <module>
    client = CRABClient()
  File "crab/3.3.1912.rc1-cms/bin/crab", line 49, in __init__
    self.subCommands = getAvailCommands()
  File "crab/3.3.1912.rc1-cms/lib/CRABClient/ClientUtilities.py", line 362, in getAvailCommands
    subcmdplugins = getPlugins(subcmdpath, subcmdname, ['SubCommand'])
  File "crab/3.3.1912.rc1-cms/lib/CRABClient/ClientUtilities.py", line 291, in getPlugins
    mod = __import__('%s.%s.%s' % (namespace, plugins, el[1]), globals(), locals(), el[1])
  File "crab/3.3.1912.rc1-cms/lib/CRABClient/Commands/getcommand.py", line 6, in <module>
    from CRABClient.UserUtilities import getMutedStatusInfo
  File "crab/3.3.1912.rc1-cms/lib/CRABClient/UserUtilities.py", line 13, in <module>
    from dbs.apis.dbsClient import DbsApi
  File "crab/3.3.1912.rc1-cms/lib/dbs/apis/dbsClient.py", line 8, in <module>
    import cjson
ImportError: No module named cjson
```

**Comment by smuzaffar** [comment]
In order to update crab in cmssw distribution one needs to request changes via PR for
- crab-prod: https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_11_1_X/master/crab.spec
- crab-pre: https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_11_1_X/master/crab-dev.spec
- crab-dev: https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_11_1_X/master/crab-dev.spec

For now all of these are using crab client `3.3.1912.rc1`

**Comment by belforte** [comment]
I can understand this:
   pycurl should comes from system too (it is available on both slc6 and slc7).

system pycurl does not have proper SSL support. Why do we build our own otherwise ?

**Comment by belforte** [comment]
we should asl Yuyi to remove cjson from dbs. I'll follow up.

**Comment by belforte** [comment]
I tested crab-prod and while other command seemed to work, submit fails
belforte@lxplus736/TC3> /cvmfs/cms-ib.cern.ch/week1/common/crab-prod submit
Will use CRAB configuration file crabConfig.py
Warning: Incompatible CRABClient version 3.3.1912.rc1
Server is saying that compatible versions are: [u'3.3.1911.rc1', u'3.3.1911', u'3.3.1911.patch1', u'3.3.1911.rc3']
Importing CMSSW configuration demoanalyzer.py
ERROR: ImportError: No module named FWCore.ParameterSet.Config

	Please email hn-cms-computing-tools@cern.ch for support with the crab.log file or crab.log URL.
	Client Version: 3.3.1912.rc1
	The log file /afs/cern.ch/work/b/belforte/CRAB3/TC3/crab_20191218_234140/crab.log has been uploaded automatically to the CRAB cache.
	Please email the following URL 'https://cmsweb.cern.ch/crabcache/logfile?name=2019-12-18_234143_crab.log&username=belforte' to hn-cms-computing-tools@cern.ch if you need help fr

**Comment by belforte** [comment]
@lecriste does it work for you ?

**Comment by smuzaffar** [comment]
@belforte , pycurl just needs to load libcurl.so. As we added curl dependency for crab, so pycurl wiill load cms libcurl (which has SSL support)

[a] System pycurl and system curl
```
lxplus> python
Python 2.7.5 (default, Aug  7 2019, 00:51:29) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-39)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import pycurl
>>> pycurl.version
'libcurl/7.29.0 NSS/3.44 zlib/1.2.7 libidn/1.28 libssh2/1.8.0'
```

[b] System pycurl and our curl
```
lxplus> source /cvmfs/cms.cern.ch/slc7_amd64_gcc820/external/curl/7.59.0/etc/profile.d/init.sh 
lxplus> python
Python 2.7.5 (default, Aug  7 2019, 00:51:29) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-39)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import pycurl
>>> pycurl.version
'libcurl/7.59.0 OpenSSL/1.0.2d zlib/1.2.8'
```

**Comment by belforte** [comment]
THanks, I think I understand the point about pycurl now.

I am fine with keeping CRAB Client compatible with sytem python,
although I must say that I was not expecting such a script which defines the env, runs, resets.
What happens to python scripts which were calling CRAB API's ? How do we run them ?

I am also a bit puzzled how I can go about testing new code, should I make a personal copy of /cvmfs/cms-ib.cern.ch/week1/common/_crab-startup and edit it to point PYTHONPATH to my git area ?

SImply stated, this looks different from what we talked about, or at least what I had in mind, and I am lost.

**Comment by smuzaffar** [comment]
We can move https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_11_1_X/master/_crab-startup.file in to crab repository and make it work in standalone mode too.

By the way, I noticed that crab needs to have cmssw PYTHONPATH available when it reads cmssw configuration. That was the reason you were not able to submit a job as mentioned here https://github.com/dmwm/CRABClient/issues/4844#issuecomment-567246077  ) .  I have proposed two changes 

- https://github.com/cms-sw/cmsdist/pull/5442/files
- https://github.com/dmwm/CRABClient/pull/4861

New crab startup script to set `CRAB_CMSSW_SITE_PACKAGES` env which CRABClient/JobType/CMSSWConfig.py can use to properly set the python before loading cmssw configuration.

With these two changes I was able to submit a test job

[a]
crab-prod submit -c ../conf/crab.py
```
Will use CRAB configuration file ../conf/crab.py
Importing CMSSW configuration /afs/cern.ch/work/m/muzaffar/crab/crab-conf/conf/pset_tutorial_analysis.py
Finished importing CMSSW configuration /afs/cern.ch/work/m/muzaffar/crab/crab-conf/conf/pset_tutorial_analysis.py
Sending the request to the server at cmsweb.cern.ch
Success: Your task has been delivered to the prod CRAB3 server.
Task name: 191219_171300:muzaffar_crab_test2_CMSSW_11_0_0_pre13_1000_step1
Project dir: crab_projects/crab_test2_CMSSW_11_0_0_pre13_1000_step1
Please use 'crab status -d crab_projects/crab_test2_CMSSW_11_0_0_pre13_1000_step1' to check how the submission process proceeds.
Log file is /afs/cern.ch/work/m/muzaffar/crab/crab-conf/run/crab_projects/crab_test2_CMSSW_11_0_0_pre13_1000_step1/crab.log
```

**Comment by smuzaffar** [comment]
About `What happens to python scripts which were calling CRAB API's ? How do we run them ?` as I have mentioned already that is bit tricky. As in this case we (crab) can not manipulate env because Python Interpreter has already been started by user processor. We can provide crab-setup script which one can source to set the crab env but there is no guarantee that every thing will work.

**Comment by belforte** [comment]
I am too thick sorry, I think I need to try things and slowly learn. Can you send me a prescription for the standalone mode ? Sort of "how do I develop and test code". 

I understand that we can not offer guarantees in case users start python in some odd env. But I thought that we can make it work when they run CRAB code inside CMSSW env, which is exactly what they do now.
Maybe I simply don't get it. I was hoping that @lecriste could follow up... otherwise it may end up that when I am at CERN in February we spend 1h together and you show me how things work ?

**Comment by smuzaffar** [comment]
@belforte , one simple way of testing crab client is to build it locally e.g.
```
> ssh lxplus
> git clone -q --depth 1 git@github.com:cms-sw/cmsdist
# if needed, edit cmsdist/crab.spec to update crab, dbs, wmcore versions to test. You can also use git commit hash instead of tag
> git clone -q --depth 1 git@github.com:cms-sw/pkgtools
> ./pkgtools/cmsBuild -a slc7_amd64_gcc820 -i build build crab
> ./build/common/crab-prod checkusername
......
.....
Retrieving username for this DN...
username is muzaffar
```

**Comment by smuzaffar** [comment]
By the way, new crab client tag `3.3.1912.rc2` is now available under /cvmfs/cms-ib.cern.ch/week1 . You should be able to test it by running `/cvmfs/cms-ib.cern.ch/week1/common/crab-prod <commands>....`

**Comment by belforte** [comment]
enlightening @smuzaffar thanks ! I am about to leave for Italy. Will try things as I can during holidays and we'll touch base again in January

**Comment by lecriste** [comment]
I ran some test and found that
- `crab submit` requires the `-c` option to specify a configuration file, used to work also without `-c`
- The `--jobids` option of `crab resubmit` is broken, I get:

> Usage: crab3bootstrap submit [options] [args]
> crab3bootstrap: error: no such option: --jobids 


About python scripts calling CRAB API's, say [this example](https://twiki.cern.ch/twiki/bin/view/CMSPublic/CRABClientLibraryAPI#Example_submitting_multiple_task), I think we need first to set PYTHONPATH in order for the script to execute, e.g.,
`from CRABClient.UserUtilities import config, getUsernameFromSiteDB`

Can this be done with `cmsenv`?

**Comment by smuzaffar** [comment]
@lecriste ,  about `crab submit`, I was able to submit without -c option [a]. 
about `resubmit`, yes its a bug in the startup script https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_11_1_X/master/_crab-startup.file#L49 . I will fix this. Basically for resubmit command it is also calling crab3bootstrap

[a]
```
> crab-prod submit
Contacting voms2.cern.ch:15002 [/DC=ch/DC=cern/OU=computers/CN=voms2.cern.ch] "cms"...
Remote VOMS server contacted succesfully.


Created proxy in /tmp/x509up_u31567.

Your proxy is valid until Sat Jan 11 15:25:22 CET 2020
Will use CRAB configuration file crabConfig.py
Warning: Incompatible CRABClient version 3.3.1912.rc2
Server is saying that compatible versions are: [u'3.3.1911.rc1', u'3.3.1911', u'3.3.1911.patch1', u'3.3.2001', u'3.3.1911.rc3']
Importing CMSSW configuration /afs/cern.ch/work/m/muzaffar/crab/crab-conf/conf/pset_tutorial_analysis.py
Finished importing CMSSW configuration /afs/cern.ch/work/m/muzaffar/crab/crab-conf/conf/pset_tutorial_analysis.py
Sending the request to the server at cmsweb.cern.ch
Success: Your task has been delivered to the prod CRAB3 server.
Task name: 200110_142544:muzaffar_crab_test2_CMSSW_11_0_0_pre13_1000_step1
Project dir: crab_projects/crab_test2_CMSSW_11_0_0_pre13_1000_step1
Please use 'crab status -d crab_projects/crab_test2_CMSSW_11_0_0_pre13_1000_step1' to check how the submission process proceeds.
Log file is /afs/cern.ch/work/m/muzaffar/crab/crab-conf/conf/crab_projects/crab_test2_CMSSW_11_0_0_pre13_1000_step1/crab.log
```

**Comment by smuzaffar** [comment]
@lecriste  https://github.com/cms-sw/cmsdist/pull/5460 should fix the `crab resubmit` issue.

**Comment by lecriste** [comment]
> @lecriste , about `crab submit`, I was able to submit without -c option [a].

@smuzaffar, the plain `crab submit` command (without an input configuration file) works for me as well. It looks for the default `crabConfig.py`:

> [a]
> 
> ```
> > crab-prod submit
> ...
> Will use CRAB configuration file crabConfig.py
> ...
> ```

If you want to use a different configuration file, `crab submit <config-file-name>` does not work any longer. It works with the `-c <config-file-name>` option though.

**Comment by smuzaffar** [comment]
For crab python API, we could deploy something like /afs/cern.ch/user/m/muzaffar/public/pycrab under /cvmfs/cms.cern.ch/share/py-crabclient and update /cvmfs/cms.cern.ch/cmsset_default.sh to include it in PYTHONPATH. This consists of two very simple python scripts which can dynamically load CRABClient/API from the latest crabclient installation. You can already try it by doign something like
```
PYTHONPATH=/afs/cern.ch/user/m/muzaffar/public/pycrab python crabConfig.py
```

**Comment by smuzaffar** [comment]
@lecriste , `crab submit <config-file-name>` is failing during `crab3bootstrap` https://github.com/dmwm/CRABClient/blob/master/bin/crab.sh#L37 . Is `crab3bootstrap`  not needed any more ?

**Comment by belforte** [comment]
@smuzaffar I believe that we are on the wrong path here and need a reset and regroup.
We went over this in CRAB meeting yesterday, but it looks like it is difficult to properly communicate.
One possibility is to wait unitl beginning of February when I'll be at CERN.
But I will also try to explain more in here. Not immediately though.

**Comment by belforte** [comment]
copying here what I wrote in CRAM meeting minutes:
```
it looks like there is a misunderstanding with Shahzad, he gave us a command which unsets scram to run crab python in the login environment, while we want an environment where we can run CRAB and CMSSW python at same time.
```
I hope we can make things much simpler by following that direction. And e.g. do things like deprecate `crab3bootstrap`. I want to simplify CRAB Client setup, not import it "as is" inside CMSSW :-)

**Comment by smuzaffar** [comment]
@belforte , my understanding was that you wanted to have a simple procedure which allows users to use latest crabclient for any cmssw .... just like scram and das client.

So just like scram and das we have to come up with a start-up crab script e.g. /cvmfs/cms-ib.cern.ch/nweek-02611/common/crab-prod which is independent of cmssw version and will always find and run latest crab client. This procedure should work for any cmssw release (even those which we have built in past).

For CRABClient python API to also work in the same way i.e. always use the latest available crabclient version, we need somehting like /afs/cern.ch/user/m/muzaffar/public/pycrab. We can make this part of cms distribution and once it is in PYTHONPATH (after sourcing cmsset_default.sh) then "import CRABClient" will always load latest crabclient version.

Frankly I can not think of any other simple procedure :-) anyway we can discuss it in Feb once you are here.

**Comment by belforte** [comment]
It is very likely that this is mostly a difficulty on my side to properly understand.
Please keep in mind that I am not the author of the current CRAB Client deployment procedure
nor ever tried to fully understand it.

I believe I understand the part about ` /afs/cern.ch/user/m/muzaffar/public/pycrab` , at least as
"somehow add to $PYTHONPATH the path CRABInstallPath" although  I would not know exactly which command will do this, but I understand `/afs/cern.ch/user/m/muzaffar/public/pycrab/CRABSetup.py`
What I fully miss is why we need anything else besides adding `CRABInstallPath/bin` to `$PATH`,
i.e. why you need to come up with a new command which wraps it.
We do not need to be able to run crab after  sourcing cmsset_default.sh, but only after having also done cmsenv.

Another way to look at things: while I think I understand e.g. `/cvmfs/cms.cern.ch/common/dasgoclient` I do not understand why we need the much more complex `/cvmfs/cms-ib.cern.ch/nweek-02611/common/crab-prod`

**Comment by smuzaffar** [comment]
I started `crab-prod`  from https://github.com/dmwm/CRABClient/blob/master/bin/crab.sh . so make  crab.sh less complex and I will update `crab-prod` :-) All crab-prod is doing is

- check which crab package to use e.g. crab-prod, crab-pre or crab-dev
   https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_11_1_X/master/_crab-startup.file#L12-L14
- find cmsos e.g slc7_amd64 and search latest crab client for this arch
  https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_11_1_X/master/_crab-startup.file#L16-L24
- unset cmssw env for non crab-submit
  https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_11_1_X/master/_crab-startup.file#L26-L29
- Set crab client runtime env
  https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_11_1_X/master/_crab-startup.file#L31-L32
- As crab clinet still have cjson dependency so if cjson if not available then make sure to include a dummy implementation. This can be removed when cjson dependency is removed
https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_11_1_X/master/_crab-startup.file#L34-L37
- Run bin/crab

Let me know hwich of these steps can be avoided.

Adding CRABInstallPath/bin to $PATH means one has to source a script which will pollute the env with crabclient dependencies which can break cmssw or other tools.

**Comment by smuzaffar** [comment]
by the way, the idea is that once you login to lxplus or a system with cms env then you do not have to source any special scripts to get/run crabclient. `crab`, `crab-pre` and `crab-dev` commands will be in  $PATH .

**Comment by belforte** [comment]
There's no need to unset CMSSW, in particular I'd rather not run CRAB with the login python.

I am leaning toward replacing DBS with dasgoclient in the only place where CRAB Client talks to DBS, to get rid of cjson dependency.

indeed crab.sh is way more complicated than I think is needed. I guess for historical reasons. Needs a good cleanup. I think I should make an attempt at that.

In the end I think that once you made CRAB Python API work (prod/pre/dev), then "crab" command should work in same env. and a simpler driving script like you have for dasgoclient should work. 

Anyhow better resume this thread after I got something working.

**Comment by smuzaffar** [comment]
I though the `gfal-*` commands will fail if we do not unset cmssw env .... has this been fixed?

**Comment by belforte** [comment]
hou are correct that this is a problem we had in the past and discussed with
you, Eric V. and David L. already at some point in the past.
It has been fixed a few years ago when you gave us `scram unsetenv` :-)

whenever CRAB Client needs to execute a Grid UI comman (voms-proxy*, gfal*, myproxy*) is
forks a suprocess where it prepends the command with scram unsetenv.

**Comment by smuzaffar** [comment]
@belforte , @lecriste can you please try the new crab-prod from /cvmfs? crab startup script is as simple as https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_11_1_X/master/crab.sh.file . 

In order to test it, please do (on lxplus, lxplus6)
```
# prepare cmssw env
export PATH=/cvmfs/cms-ib.cern.ch/week1/common:${PATH:+$PATH}
export PYTHONPATH="/cvmfs/cms-ib.cern.ch/week1/share/cms/crab-common/1.0/lib${PYTHONPATH:+:$PYTHONPATH}"
# now you should be able to run crab commands and use CRAB Python API
```

**Comment by belforte** [comment]
thanks @smuzaffar ! I am using `crab-prod` now, looks good. so far I will test more things.

**Comment by belforte** [comment]
command completion by TAB is not working.

**Comment by belforte** [comment]
P.S. we may need to slightly modify `CRABClient/etc/crab-bash-completion.sh` to make it work also with crab-pre and crab-dev, at least IIUC this line
```
complete -F _UseCrab -o filenames crab
```

**Comment by belforte** [comment]
currently TAB completion works after
```
source /afs/cern.ch/user/b/belforte/WORK/CRAB3/CRABClient/etc/crab-bash-completion.sh
```
I tried several CRAB Client command and they work, I see little point in testing all of them. Things which @lecriste found earlier (like need to use -c config, or problems with submit) are not there anymore.

**Comment by belforte** [comment]
I tested also CRAB API and it is OK.
Thanks Shahzad and Leonardo.
What is next step  ?

**Comment by belforte** [comment]
@smuzaffar I tagged CRAB.3.3.2001.patch1 with a fix to the command autocompletion, if you can make this file [1] appear somewhere in /cvmfs/cms-ib.cern.ch/...  that it can be sourced as part of crab setup.
[1]
https://github.com/dmwm/CRABClient/blob/master/etc/crab-bash-completion.sh

Here's the setup script which I am using no, showting the changes.
@smuzaffar am I correct in believing that eventually the two exports will not be needed anymore ?
```
#!/bin/bash

export SCRAM_ARCH=slc7_amd64_gcc700
cd /afs/cern.ch/user/b/belforte/WORK/CMSSW/SL7/CMSSW_10_2_6/

eval `scramv1 runtime -sh`

# use current Client distribution
# source /cvmfs/cms.cern.ch/crab3/crab.sh

# use Client distribution inside CMSSW, from Shahzad
export PATH=/cvmfs/cms-ib.cern.ch/week1/common:${PATH:+$PATH}
export PYTHONPATH="/cvmfs/cms-ib.cern.ch/week1/share/cms/crab-common/1.0/lib${PYTHONPATH:+:$PYTHONPATH}"
# temporary
alias crab=crab-prod
source /afs/cern.ch/user/b/belforte/WORK/CRAB3/CRABClient/etc/crab-bash-completion.sh

# now you should be able to run crab commands and use CRAB Python API

cd /afs/cern.ch/user/b/belforte/WORK/CRAB3/TC3

```

**Comment by smuzaffar** [comment]
@belforte , you asked for next steps.
- `export PATH=/cvmfs/cms-ib.cern.ch/week1/common:${PATH:+$PATH}` part will not be needed once we have a cmssw release build and installed with new `crab-*` packages. Currently default `crab-*` commands (i.e. you login to lxplus and  run `crab-*` without sourcing any setup script) comes from `/cvmfs/cms.cern.ch` which is an old version crab client.
- setting up `PYTHONPATH` can also be automatic. `cmsset_default.*` scripts allow us to source various startup scripts. This means once you login to lxplus then `/cvmfs/cms.cern.ch/share/cms/crab-common/1.0/lib` can be there in PYTHONPATH. I am not sure if this will break any thing. `/cvmfs/cms.cern.ch/share/cms/crab-common/1.0/lib`  only contains `CRABAPI` and `CRABClient` pythion modules, so ideally this should not break any thing. But if we want to be on safe side then we can ask users to source a script (just like they do it now) to get crab client in PYTHONPATH. 
Let me know if you prefer the automatic setup or would you prefer that users always source some crab-setup script?
- sourcing of `etc/crab-bash-completion.sh` has been fixed and will be part of next update. This should work for all `crab-*` commands.

**Comment by smuzaffar** [comment]
@belforte , we also need to fix the versioning scheme. Currently there is no easy way to find which version of crabclient is latest for example if we have following versions installed 
- 3.3.2001
- 3.3.2001.rc1
- 3.3.2001.rc2
- 3.3.2001.patch1
- 3.3.2001.patch2
then its hard to find which version is newer than other.

**Comment by lecriste** [comment]
For a given year-month tag (`3.3.yymm`) the chronological order is
1. `3.3.yymm.rc**x**
2. `3.3.yymm`
3. `3.3.yymm.patch**y** 

where **x** and **y** follow the standard numerical order.

**Comment by smuzaffar** [comment]
how you programatically find the latest version? I need to implement is in `bash` so that `crab-*` can correctly use the latest version installed.

**Comment by lecriste** [comment]
I have not a script, we can change the versioning schema as you wish.
By the way, the `rc`**x** tags are intended for `crab-pre`, the other two tag types for `crab` command.
Perhaps it is easier to find the latest tags in these two cases.

**Comment by belforte** [comment]
@smuzaffar if there's a different standard that you are using and/or you'd like to suggest, we can surely adopt it and make things easier for everybody, current scheme is already an attempt to follow what others are doing, but indeed is more geared to human understanding than machine one

**Comment by belforte** [comment]
All in all "starting with 2020" is a good point to change naming standard. We could e.g. drop the `3.3` which carries no info

**Comment by belforte** [comment]
@smuzaffar see answers/comments inline
> @belforte , you asked for next steps.
> 
>     * `export PATH=/cvmfs/cms-ib.cern.ch/week1/common:${PATH:+$PATH}` part will not be needed once we have a cmssw release build and installed with new `crab-*` packages. Currently default `crab-*` commands (i.e. you login to lxplus and  run `crab-*` without sourcing any setup script) comes from `/cvmfs/cms.cern.ch` which is an old version crab client.
> 
OK. But surely it will be good to remove those old commands (which I did not even know existed !), who can do this ?

>     * setting up `PYTHONPATH` can also be automatic. `cmsset_default.*` scripts allow us to source various startup scripts. This means once you login to lxplus then `/cvmfs/cms.cern.ch/share/cms/crab-common/1.0/lib` can be there in PYTHONPATH. I am not sure if this will break any thing. `/cvmfs/cms.cern.ch/share/cms/crab-common/1.0/lib`  only contains `CRABAPI` and `CRABClient` pythion modules, so ideally this should not break any thing. But if we want to be on safe side then we can ask users to source a script (just like they do it now) to get crab client in PYTHONPATH.
>       Let me know if you prefer the automatic setup or would you prefer that users always source some crab-setup script?
Let's stay with an explicit crab-setup for a while. In a few months, when user have transitioned and
we are confident there is no problem (even if we do not expect those)  we can change.
> 
>     * sourcing of `etc/crab-bash-completion.sh` has been fixed and will be part of next update. This should work for all `crab-*` commands.

Thanks !

**Comment by smuzaffar** [comment]
@belforte , the old commands are part of new distribution mechanism and were installed as a part of CMSSW_11_1_0_pre1 release. New crab version (and its `crab-*` commands) will be installed in default /cvmfs/cms.cern.ch once we deploy a new cmssw version.

**Comment by belforte** [comment]
@smuzaffar as I see that you are going ahead with more work here, let me ask before you finalize next release:
 when is the good time to e.g. change crab-pre.spec or crab-dev.spec to use new CRABClient tag ?

**Comment by smuzaffar** [comment]
@belforte , I think I am mostly done with the changes. I was just waiting for CMSSW 11.1.0.pre2 to be out so that it can deploy the new crab-* versions.
Feel free to update the tag in https://github.com/cms-sw/cmsdist/pull/5479 for crab-pre|dev now. If we manage to merge it today then it might end up in cmssw 11.1.0.pre2 release.

**Comment by lecriste** [comment]
@smuzaffar do you know if anything changed in
`/cvmfs/cms.cern.ch/crab3/crab.sh`
?
See:
https://hypernews.cern.ch/HyperNews/CMS/get/computing-tools/5490/1/1.html

**Comment by smuzaffar** [comment]
No nothing changed due to new crab. looks like /cvmfs/cms.cern.ch/crab3/crab_standalone.sh is missing or env CRAB_SOURCE_SCRIPT is not set properly before sourcing the crab,sh. I have no idea who creates/deploy this file
```
bash -ex  /cvmfs/cms.cern.ch/crab3/crab.sh
+ '[' -z '' ']'
+ CRAB_SOURCE_SCRIPT=/cvmfs/cms.cern.ch/crab3/crab_standalone.sh
++ getVariableValue '$CRABCLIENT_ROOT' bin
++ VARNAME='$CRABCLIENT_ROOT'
++ SUBDIR=bin
++ sh -c 'source /cvmfs/cms.cern.ch/crab3/crab_standalone.sh >/dev/null 2>/dev/null; if [ $? -eq 0 ] && [ -d $CRABCLIENT_ROOT ];                     then echo $CRABCLIENT_ROOT/bin; else exit 1; fi'
+ CRAB3_BIN_ROOT=

```

**Comment by smuzaffar** [comment]
looks like there are broken symlinks
```
lrwxrwxrwx. 1 cvmfs cvmfs    85 Dec 28 21:07 crab.csh -> /cvmfs/cms.cern.ch/crab3/slc7_amd64_gcc630/cms/crabclient/3.3.2001/etc/init-light.csh
lrwxrwxrwx. 1 cvmfs cvmfs    93 Dec 27 15:07 crab_pre.csh -> /cvmfs/cms.cern.ch/crab3/slc7_amd64_gcc630/cms/crabclient/3.3.2001.rc1/etc/init-light-pre.csh
lrwxrwxrwx. 1 cvmfs cvmfs    92 Dec 27 15:07 crab_pre.sh -> /cvmfs/cms.cern.ch/crab3/slc7_amd64_gcc630/cms/crabclient/3.3.2001.rc1/etc/init-light-pre.sh
lrwxrwxrwx. 1 cvmfs cvmfs    93 Jan 22 11:41 crab_pre_standalone.csh -> /cvmfs/cms.cern.ch/crab3/slc7_amd64_gcc630/cms/crabclient/3.3.2002.rc1/etc/profile.d/init.csh
lrwxrwxrwx. 1 cvmfs cvmfs    92 Jan 22 11:41 crab_pre_standalone.sh -> /cvmfs/cms.cern.ch/crab3/slc7_amd64_gcc630/cms/crabclient/3.3.2002.rc1/etc/profile.d/init.sh
lrwxrwxrwx. 1 cvmfs cvmfs    84 Dec 28 21:07 crab.sh -> /cvmfs/cms.cern.ch/crab3/slc7_amd64_gcc630/cms/crabclient/3.3.2001/etc/init-light.sh

```

**Comment by lecriste** [comment]
Ok, hopefully Bockjoo con help there.
By the way is this
https://hypernews.cern.ch/HyperNews/CMS/get/computing-tools/5489/1/2.html
expected?

**Comment by belforte** [comment]
@smuzaffar @amaltaro I have found a small sticky point.
CRABClient still has a corner case were it looks up DBS using WMCore DBSReader. As discussed elsewhere, that now requires retry decorator [1] which
AFAICT is no part of CMSSW python [2]
In COMP side where services like CRAB TaskWorker run in their own python built, we add py2-retry to the spec file [3]. But we always run CRAB Client with CMSSW python.

I see 3 ways out and would like your opinion on which to choose or to suggest a better one
1. change CRABClient to use DASGOCLIENT instead of DBS (good thing anyhow, but takes a bit of my time)
2. Shahzad adds py2-retry to CMSSW's python (maybe it can be useful elsewhere as well)
3. There is some other funciton/method/decorator in CMSSW python which achieves same result and Alan can try instead

Kindly let me know. I think we can build CRABClient with old WMCore 1.2.8 until this is sorted out, but should not get locked  out of WMCore updates.

[1]
https://pypi.org/project/retry/
[2]
```
belforte@lxplus754/cmsdist> which python
/cvmfs/cms.cern.ch/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_2_6/external/slc7_amd64_gcc700/bin/python
belforte@lxplus754/cmsdist> python
Python 2.7.14+ (default, Mar 28 2018, 11:00:22) 
[GCC 7.3.1 20180127] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import retry
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: No module named retry
>>> 
```
[3]
https://github.com/cms-sw/cmsdist/blob/eff210dc14ebb01e9f0f0291c3ef84cc79bf64ac/crabtaskworker.spec#L16

**Comment by belforte** [comment]
Currently my preference is for 3. so to give me time to do 2. As to going to dasgoclient... it is appealing, but of all tools around DBS Python API is the one with the best documentation IMHO, so it looks faster. The only drawback wrt using dasgoclient is that cjson is needed, but it is there already.

**Comment by smuzaffar** [comment]
opts 3 and 4 are not going to fix existing cmssw versions. Better to go for 1

**Comment by amaltaro** [comment]
I'm not very fond of it, but another possibility would be to actually move that function `getDataTiers` with the retry decorator to a different place. And we leave the `getDataTiers` without the decorator inside the DBSReader module. Please let me know what you think, and I can make that change.

**Comment by smuzaffar** [comment]
By the way, new crabclient 3.3.2001 has been distributed via CMSSW 11.1.0.pre2 and is available on lxplus (under `/cvmfs/cms.cern.ch`). New `crab-prod|dev|pre` commands are now available for all archs (slc6, slc7 and cc8) on lxplus. One should be able to use these commands without sourcing any extra script (of course these commands will only work under cmssw env).

In order to use crabclient python API, one needs to `source /cvmfs/cms.cern.ch/common/crab-setup.[c]sh` . It just adds crab clients lib's directory in to `PYTHONPATH` . By default it sets crab `prod` but one can `source /cvmfs/cms.cern.ch/common/crab-setup.[c]sh dev|pre` to get `pre` or `dev` versions of crab client.

Can you please give it a try on lxplus. Again this is still under tests so please do not recommend these `crab-prod|pre|dev` commands to users.

**Comment by belforte** [comment]
thanks. Yes, I will test carefully and get back to you. I will also make a PR to roll back WMCore from 1.2.9 to 1.2.8 until I have changed the code in order not to need py2-retry. I will implemento either 1. or 2. in my list. No action needed from @smuzaffar nor @amaltaro

**Comment by belforte** [comment]
@smuzaffar crab commands autocompletion does not work. I guess this line
`complete -F _UseCrab_${CRABCLIENT_TYPE} -o filenames crab`
https://github.com/cms-sw/cmsdist/blob/4b38b5c8ef9fde034348f1570e6ffeb8d2fc1786/crab/crab-setup.sh.file#L15
which appears now in `/cvmfs/cms.cern.ch/common/crab-setup.sh` should rather be in `/cvmfs/cms.cern.ch/common/crab-prod|pre|dev` . Well.. it is harmless in the `crab-setup.sh`, simply un-needed for using python API. I do not understand how the `crab-*` commands are create, so do not know which change is needed. In case, it is surely fine to execute it at all times, whether one is going to use python API or not.

Of course I keep testing by sourcing myself the `complete` command.

**Comment by belforte** [comment]
@smuzaffar aside from "it works", about the overall design. I.e. how it is supposed to work.

I do not see it as user-friendly that one has to surce some script only when using the python API. Is this because of the worry that adding CRAB stuff to PYTHONPATH may break something in CMSSW ?  In any case it may easily lead to confusion on user sides at at minimum a lot of run/fail/oops/setup/runagain/swear-and-curse if not complain.
Did I manage to confuse you one more time with some previous message ?

I would find it simpler if we start like now : user needs to source something after "cmsenv" to use CRAB (also avoids that they source current setup on top of what is in CMSSW). Then if/when you are confident that CRAB does not break anything, we can make it part of cmsenv and dummyfy the setup script.

Or am I confused and missing some important fact ?

**Comment by smuzaffar** [comment]
@belforte , looks like new cms-common packages was not installed for /cvmfs/cms.cern.ch . cms-common is a special package and needs special treatment from Bockjoo. I am going to ask him to update it.

New cms-common should automatically source the autocomplete scripts for crab-* commands (/cvmfs/cms.cern.ch/share/etc/profile.d/S99crab-env.sh ).

**Comment by smuzaffar** [comment]
No, you did not confuse me. Yes, for users the instructions will be that they have to source some setup scripts.

But during this test phase, I wanted to check the proof of concept that we can run command-line crab client without sourcing anything. For now sourcing /cvmfs/cms.cern.ch/common/crab-setup.sh only updates PYTHONPATH that is why I asked to source it (for your testing) for python API tests. But there is no harm to source it always.

**Comment by belforte** [comment]
ThanksSent from. Stefano's phone

**Comment by belforte** [comment]
@smuzaffar  Bockjoo updated cvmfs and now the command completion works.
(even if I could not figure out how it is done!)
But I noticed a discrepancy:
when using e.g. `crab_prod`, python will run with this $PYTHONPATH
`/cvmfs/cms.cern.ch/share/cms/crab-prod/3.3.2001/lib`
while after` source /cvmfs/cms.cern.ch/common/crab-setup.sh` it is set to
`/cvmfs/cms.cern.ch/share/cms/crab/1.0/lib/prod`

I presume this is intentional, even if a bit confusing.
So I have "discovered" that after sourcing `crab_setup.sh` I need to do
`import CRABClient`
in python, before I can do
`from dbs.apis.dbsClient import DbsApi`
If that is going to stay, we'll have to revisit a bit the documentation.


I am indeed intrigued by the magic in `/cvmfs/cms.cern.ch/share/cms/crab/1.0/lib/crab-proxy-package`.  But since I was not expecting it, but imagining a more old fashioned $PYTHONPATH pointing to a `lib` directory with all dependencies in , like `/cvmfs/cms.cern.ch/share/cms/crab-prod/3.3.2001/lib`, I am now puzzled at how to make it possible to use DBS3 scripts like those pointed to in https://twiki.cern.ch/twiki/bin/view/CMSPublic/Crab3DataHandling#Changing_a_dataset_or_file_statu

 was prepared to tell users to fetch them gitHub, or to clone them to CRABClient/bin or whatever (they are extremely stable), but how do we run them ? Somehow inserting `import CRABClient` in there does not look good. 
I agree that this is going beyond the scope "make crab command work inside CMSW", my fault, I was wrongly assuming that it was going to come for free.

**Comment by smuzaffar** [comment]
@belforte , `cmsset_default.sh` script, which is sourced when one login, had the functionality to source addtional group startup scripts ( https://github.com/cms-sw/cms-common/blob/master/cmsset_default.sh#L24-L33 ) . I made use of it and created `/cvmfs/cms.cern.ch/share/etc/profile.d/S99crab-env.sh` .   `S99crab-env.sh` sources the `crab-bash-completion.sh`.

Yes the python path is intentional, `crab-*` command-line clients use lib directly from the latest crabclient installation. This is done as crab-* client do not pollute env and they can set PYTHONPATH to `/cvmfs/cms.cern.ch/share/cms/crab-prod/3.3.2001/lib` where we have few python packages  e.g. dbs,  Utils etc.  While `/cvmfs/cms.cern.ch/common/crab-setup.sh` pollute env (by setting PYTHONPATH) so I did not want to have these `dbs, Utils` modules available by default. This can break is env of any other package needs `dbs` but a different version which is not competible with version in crab.

For the same reason above, I only made CRABAPI and CRABClient available via PYTHONPATH (after sourcing crab-setup.sh). This distribution is suppose to provide CRAB* API and not dbs. So user should be import CRAB* first :-)

**Comment by belforte** [comment]
I understand. Thanks.
I will take care of  DBS3SetDatasetStatus.py etc. via proper instructions in the twiki.
Hopefully we'll switch user produced files to Rucio this year and that will be obsolete.

**Comment by belforte** [comment]
I have updated the spec for dev with the new CRABClient tag which should work with new WMCore.\https://github.com/cms-sw/cmsdist/pull/5501
What's the procedure for testing that ? Will it appear automatically as crab-dev ?

**Comment by smuzaffar** [comment]
I have started the test. For now tst only make usre that we can built it. If you can provide us a simple `bash` script to test few commands then we can include that in cmssw to make sure we can run new version of crab. shell script could just run few crab commands e.g.
```
crab-dev --version
crab-dev checkusername
```
you an also provide a python script to test CRAB Python API functionality.

By the way, why does crab always create crab.log file? Sometimes I run it from cvmfs direcotry and it fails to create this log file [a].

[a]
IOError: [Errno 30] Read-only file system: '/cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_0_pre2/crab.log'

**Comment by belforte** [comment]
thanks. Testing will hopefully come up care of our new operator.
About the log.. well it has been like this since ever, we try to keep
stdout terse but when someone reports a problem the full log is needed.
I do not know if there's a way to avoid it, but if useful it can be added,
I will have a look.
In any case things appear to work quite well. I'd like to understand
the process now: how/when get things deployed, what is automatic and
what require action, and again "what's the best way to use my
git clone in place of the libraries in CVMFS for developing",
and how do we roll back in case we screw up.

**Comment by belforte** [comment]
there is no option in CRABCLient to disable the log file. Let me know if you need it. I think it could be rather easy via an env.var, otherwise I have to learn the part of CRAB code which parses command arguments and add an option.

**Comment by belforte** [comment]
Everything is not working satisfactorely in my opinion, ad the new procedures to change, deploy and use CRAB Client are documented in the twiki
https://twiki.cern.ch/twiki/bin/view/CMSPublic/CMSCrabClient

Latest CRAB Client works with latest WMCore (1.2.9) and in crab report compares result from DBS Python API and DASGOCLIENT. When I ill be sure of my handling of dasgoclient, we can stop usig DBS API i the client, but keep as dependency for distribution.

Thanks all
